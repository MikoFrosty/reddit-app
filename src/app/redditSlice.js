import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";

const initialState = {
  posts: [],
  comments: [],
  status: "idle",
  commentsStatus: "idle",
  commentsId: "",
  subreddit: "top",
  searchTerm: "",
};

export const fetchReddit = createAsyncThunk(
  "reddit/fetchPosts",
  async (subreddit = "home") => {
    const response = await fetch(`https://www.reddit.com/${subreddit}.json`);
    const json = await response.json();
    const posts = json.data.children.map((child) => child.data);
    return posts;
  }
);

export const fetchComments = createAsyncThunk(
  "reddit/fetchComments",
  async (permalink) => {
    const response = await fetch(`https://www.reddit.com${permalink}.json`);
    const json = await response.json();
    const comments = json[1].data.children.map((child) => child.data);
    return comments;
  }
);

export const redditSlice = createSlice({
  name: "reddit",
  initialState,
  reducers: {
    setCommentsId: (state, action) => {
        state.commentsId = action.payload;
        },
    setSubreddit: (state, action) => {
      state.subreddit = action.payload;
    },
    setSearchTerm: (state, action) => {
      state.searchTerm = action.payload;
    },
  },
  extraReducers: {
    [fetchReddit.pending]: (state) => {
      state.status = "loading";
      state.searchTerm = "";
    },
    [fetchReddit.fulfilled]: (state, action) => {
      state.status = "idle";
      state.posts = action.payload;
    },
    [fetchReddit.rejected]: (state, action) => {
      state.status = "rejected";
      state.posts = [];
    },
    [fetchComments.pending]: (state) => {
      state.commentsStatus = "loading";
    },
    [fetchComments.fulfilled]: (state, action) => {
      state.commentsStatus = "idle";
      state.comments = action.payload;
    },
    [fetchComments.rejected]: (state, action) => {
      state.commentsStatus = "rejected";
      state.comments = [];
    },
  },
});

export const selectRedditStatus = (state) => state.reddit.status;
export const selectRedditResults = (state) => state.reddit.posts;
export const selectComments = (state) => state.reddit.comments;
export const selectCommentsId = (state) => state.reddit.commentsId;
export const selectCommentsStatus = (state) => state.reddit.commentsStatus;
export const selectSubreddit = (state) => state.reddit.subreddit;
export const selectSearchTerm = (state) => state.reddit.searchTerm;

export default redditSlice.reducer;
export const { setCommentsId, setSubreddit, setSearchTerm } = redditSlice.actions;
