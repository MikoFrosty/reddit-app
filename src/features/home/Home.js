import Skeleton from "react-loading-skeleton";
import Card from "../card/Card";
import { useSelector, useDispatch } from "react-redux";
import {
  selectRedditResults,
  fetchReddit,
  selectRedditStatus,
} from "../../app/redditSlice";
import { useEffect, useState } from "react";

export default function Home() {
  const [subreddit, setSubreddit] = useState("top");
  const status = useSelector(selectRedditStatus);
  // remember when setting new subreddits to prefix with r/

  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(fetchReddit(subreddit));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [subreddit]);
  const redditResults = useSelector(selectRedditResults);

  const [page, setPage] = useState(1);
  const [postsPerPage] = useState(10);
  const indexOfLastPost = page * postsPerPage;
  //const indexOfFirstPost = indexOfLastPost - postsPerPage; // not used
  const currentPosts = redditResults.slice(0, indexOfLastPost);
  //const paginate = (pageNumber) => setPage(pageNumber); // not used

  const homeContainerStyle = {
    width: "100%",
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center",
  };

  function handleLoadClick() {
    setPage(page + 1);
  }

  if (status === "loading") {
    return (
      <div style={homeContainerStyle}>
        <Skeleton count={5} containerClassName="skeleton-container" height="100px"/>
      </div>
    );
  }

  return (
    <div style={homeContainerStyle}>
      {currentPosts.map((result) => (
        <Card key={result.id} result={result} />
      ))}
      {redditResults.length > currentPosts.length ? (
        <button onClick={handleLoadClick}>Load More</button>
      ) : (
        <p style={{ fontWeight: "bold" }}>End of Feed</p>
      )}
    </div>
  );
}
