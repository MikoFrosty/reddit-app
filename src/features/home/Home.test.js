import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../app/store";
import Home from "./Home";
import renderer from "react-test-renderer";

/////// TESTING MOCKS ///////
/*
// mock the fetchReddit action creator
const fetchReddit = jest.mock("../../app/redditSlice", () => ({
  ...jest.requireActual("../../app/redditSlice"),
  fetchReddit: jest.fn(),
}));

// mock the useSelector hook
const useSelector = jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: jest.fn(),
})); 

const useDispatch = jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useDispatch: jest.fn(),
}));
const useEffect = jest.mock("react", () => ({
    ...jest.requireActual("react"),
    useEffect: jest.fn(),
}));
const useState = jest.mock("react", () => ({
    ...jest.requireActual("react"),
    useState: jest.fn(),
}));
// mock selectSubreddit
const selectSubreddit = jest.mock("../../app/redditSlice", () => ({
    ...jest.requireActual("../../app/redditSlice"),
    selectSubreddit: jest.fn(),
}));
// mock selectRedditStatus
const selectRedditStatus = jest.mock("../../app/redditSlice", () => ({
    ...jest.requireActual("../../app/redditSlice"),
    selectRedditStatus: jest.fn(),
}));
// mock selectRedditResults
const selectRedditResults = jest.mock("../../app/redditSlice", () => ({
    ...jest.requireActual("../../app/redditSlice"),
    selectRedditResults: jest.fn(),
}));
// mock selectSearchTerm
const selectSearchTerm = jest.mock("../../app/redditSlice", () => ({
    ...jest.requireActual("../../app/redditSlice"),
    selectSearchTerm: jest.fn(),
}));
*/

describe("Home", () => {
  beforeEach(() => {
    // IntersectionObserver isn't available in test environment
    const mockIntersectionObserver = jest.fn();
    mockIntersectionObserver.mockReturnValue({
      observe: () => null,
      unobserve: () => null,
      disconnect: () => null,
    });
    window.IntersectionObserver = mockIntersectionObserver;
    // fix for video pause() error
    jest
      .spyOn(window.HTMLMediaElement.prototype, "pause")
      .mockImplementation(() => {});
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  test("matches snapshot after 100ms (store status should be 'loading')", async () => {
    render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    // wait 100ms
    await waitFor(() => new Promise((r) => setTimeout(r, 100)));

    const tree = renderer
      .create(
        <Provider store={store}>
          <Home />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders without crashing", () => {
    render(
      <Provider store={store}>
        <Home />
      </Provider>
    );
  });

  it("renders skeletons on page load", async () => {
    render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    await waitFor(() => {
      expect(
        // eslint-disable-next-line testing-library/no-node-access
        document.querySelector(".react-loading-skeleton")
      ).toBeInTheDocument();
    });
  });

  it("render search field after loading posts (3 second wait)", async () => {
    render(
      <Provider store={store}>
        <Home />
      </Provider>
    );
    await waitFor(
      () => {
        expect(
          screen.getByPlaceholderText("Search Posts...")
        ).toBeInTheDocument();
      },
      {
        timeout: 3000,
      }
    );
  });

  it("renders posts after fetch (3 second wait)", async () => {
    render(
      <Provider store={store}>
        <Home />
      </Provider>
    );
    await waitFor(
      () => {
        const container = screen.getByTestId("redditResults");
        expect(container).toBeInTheDocument();
      },
      {
        timeout: 3000,
      }
    );
    await waitFor(() => {
      screen.findAllByTestId("redditPost").then((posts) => {
        expect(posts).toHaveLength(10);
      });
    });
  });

  it("renders a load more button after loading posts (3 second wait)", async () => {
    render(
      <Provider store={store}>
        <Home />
      </Provider>
    );
    await waitFor(
      () => {
        expect(screen.getByText("Load More")).toBeInTheDocument();
      },
      {
        timeout: 3000,
      }
    );
    await waitFor(() => {
      expect(screen.getByText("Load More")).toBeEnabled();
    });
    await waitFor(() => {
      expect(screen.getByText("Load More")).toHaveAttribute("type", "button");
    });
  });
});
