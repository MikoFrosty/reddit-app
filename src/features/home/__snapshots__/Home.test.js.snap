// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches snapshot after 100ms (store status should be 'loading') 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "color": "#fff",
      "display": "flex",
      "flexDirection": "column",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <span
    aria-busy={true}
    aria-live="polite"
    className="skeleton-container"
  >
    <span
      className="react-loading-skeleton"
      style={
        Object {
          "--base-color": "#252525",
          "--highlight-color": "#505050",
          "border": "1px solid #505050",
          "height": "100px",
          "marginBottom": 10,
        }
      }
    >
      ‌
    </span>
    <br />
    <span
      className="react-loading-skeleton"
      style={
        Object {
          "--base-color": "#252525",
          "--highlight-color": "#505050",
          "border": "1px solid #505050",
          "height": "100px",
          "marginBottom": 10,
        }
      }
    >
      ‌
    </span>
    <br />
    <span
      className="react-loading-skeleton"
      style={
        Object {
          "--base-color": "#252525",
          "--highlight-color": "#505050",
          "border": "1px solid #505050",
          "height": "100px",
          "marginBottom": 10,
        }
      }
    >
      ‌
    </span>
    <br />
    <span
      className="react-loading-skeleton"
      style={
        Object {
          "--base-color": "#252525",
          "--highlight-color": "#505050",
          "border": "1px solid #505050",
          "height": "100px",
          "marginBottom": 10,
        }
      }
    >
      ‌
    </span>
    <br />
    <span
      className="react-loading-skeleton"
      style={
        Object {
          "--base-color": "#252525",
          "--highlight-color": "#505050",
          "border": "1px solid #505050",
          "height": "100px",
          "marginBottom": 10,
        }
      }
    >
      ‌
    </span>
    <br />
  </span>
</div>
`;
